File: paleo_workbench/ui/pages/map_edit_snap.py
from __future__ import annotations

from typing import Any, Iterable

from paleo_workbench.mapping import map_edit_api as api
from paleo_workbench.ui.pages.map_edit_items import (
    FaciesPolygonItem,
    LabelItem,
    LineItem,
    WellPointItem,
)

_DEFAULT_SNAP_TOL = 0.5


class MapSnapManager:
    """Manages snap state, candidate point extraction, caching, and point snapping."""

    def __init__(self, snap_tolerance: float = _DEFAULT_SNAP_TOL) -> None:
        self._snap_enabled = False
        self._snap_tolerance = snap_tolerance
        self._snap_candidate_cache: list[tuple[float, float]] | None = None
        self._snap_candidate_builds = 0
        self._reference_snap_points: list[tuple[float, float]] = []

    @property
    def enabled(self) -> bool:
        return self._snap_enabled

    @enabled.setter
    def enabled(self, value: bool) -> None:
        self._snap_enabled = bool(value)

    @property
    def tolerance(self) -> float:
        return self._snap_tolerance

    @tolerance.setter
    def tolerance(self, value: float) -> None:
        self._snap_tolerance = max(0.0, float(value))

    @property
    def reference_points(self) -> list[tuple[float, float]]:
        return self._reference_snap_points

    def set_reference_points(self, points: list[tuple[float, float]]) -> None:
        self._reference_snap_points = [(float(x), float(y)) for x, y in points]
        self.invalidate_candidates()

    def invalidate_candidates(self) -> None:
        self._snap_candidate_cache = None

    def build_count(self) -> int:
        return self._snap_candidate_builds

    def snap_xy(
        self,
        x: float,
        y: float,
        items: Iterable[Any],
        is_layer_visible_fn,
        draft_points: list[list[float]] | None = None,
    ) -> tuple[float, float]:
        if not self._snap_enabled:
            return float(x), float(y)
        candidates = self.get_candidates(items, is_layer_visible_fn, draft_points)
        return api.snap_point(candidates, float(x), float(y), tol=self._snap_tolerance)

    def get_candidates(
        self,
        items: Iterable[Any],
        is_layer_visible_fn,
        draft_points: list[list[float]] | None = None,
    ) -> list[tuple[float, float]]:
        extra = [tuple(p) for p in draft_points] if draft_points else []
        if self._snap_candidate_cache is not None:
            return [*self._snap_candidate_cache, *extra]

        pts: list[tuple[float, float]] = []
        for item in items:
            if not is_layer_visible_fn(getattr(item, "kind", "")):
                continue
            if isinstance(item, FaciesPolygonItem):
                for ring in item.all_rings():
                    for p in ring:
                        pts.append((float(p[0]), float(p[1])))
            elif isinstance(item, LineItem):
                for p in item.coordinates():
                    pts.append((float(p[0]), float(p[1])))
            elif isinstance(item, (WellPointItem, LabelItem)):
                rec = item.to_record()
                c = rec.get("coordinates") or [0, 0]
                pts.append((float(c[0]), float(c[1])))
        pts.extend(self._reference_snap_points)
        self._snap_candidate_cache = pts
        self._snap_candidate_builds += 1
        return [*pts, *extra]

File: libs/ui_data_core/src/map_edit_snap.cpp
#include "pwb/ui_data_core/map_edit_snap.hpp"

#include <type_traits>
#include <utility>

namespace pwb::ui_data_core {

void MapSnapManager::set_reference_points(std::vector<MapPoint> points) {
    reference_snap_points_ = std::move(points);
    invalidate_candidates();
}

MapPoint MapSnapManager::snap_xy(
    double x, double y, const std::vector<SnapItem>& items,
    const std::function<bool(std::string_view)>& is_layer_visible,
    const std::vector<MapPoint>* draft_points) {
    if (!snap_enabled_) {
        return {x, y};
    }
    const std::vector<MapPoint> candidates =
        get_candidates(items, is_layer_visible, draft_points);
    return snap_point(candidates, x, y, snap_tolerance_);
}

std::vector<MapPoint> MapSnapManager::get_candidates(
    const std::vector<SnapItem>& items,
    const std::function<bool(std::string_view)>& is_layer_visible,
    const std::vector<MapPoint>* draft_points) {
    std::vector<MapPoint> extra;
    if (draft_points != nullptr) {
        extra = *draft_points;
    }
    if (snap_candidate_cache_.has_value()) {
        std::vector<MapPoint> out = *snap_candidate_cache_;
        out.insert(out.end(), extra.begin(), extra.end());
        return out;
    }

    std::vector<MapPoint> pts;
    for (const SnapItem& item : items) {
        const std::string_view kind = std::visit(
            [](const auto& model) -> std::string_view {
                return model != nullptr ? std::string_view(model->kind)
                                        : std::string_view{};
            },
            item);
        if (is_layer_visible && !is_layer_visible(kind)) {
            continue;
        }
        std::visit(
            [&](const auto& model) {
                if (model == nullptr) {
                    return;
                }
                using Model = std::decay_t<decltype(*model)>;
                if constexpr (std::is_same_v<Model, FaciesPolygonModel>) {
                    for (const auto& ring : model->all_rings()) {
                        for (const auto& p : ring) {
                            pts.push_back(p);
                        }
                    }
                } else if constexpr (std::is_same_v<Model, LineModel>) {
                    for (const auto& p : model->coordinates()) {
                        pts.push_back(p);
                    }
                } else {
                    // Well / Label: rec.get("coordinates") or [0, 0].
                    pts.push_back({model->x, model->y});
                }
            },
            item);
    }
    pts.insert(pts.end(), reference_snap_points_.begin(),
               reference_snap_points_.end());
    snap_candidate_cache_ = pts;
    snap_candidate_builds_ += 1;
    std::vector<MapPoint> out = std::move(pts);
    out.insert(out.end(), extra.begin(), extra.end());
    return out;
}

}  // namespace pwb::ui_data_core
